lib/internal/modules/esm/resolve.js

```diff
@@ -132,13 +134,36 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
 
 const realpathCache = new SafeMap();
 
-/**
- * @param {string | URL} url
- * @returns {boolean}
- */
-function fileExists(url) {
-  return internalModuleStat(toNamespacedPath(toPathIfFileURL(url))) === 0;
-}
+const mainResolveExtensions = [
+  '',
+  '.js',
+  '.json',
+  '.node',
+  '/index.js',
+  '/index.json',
+  '/index.node',
+  './index.js',
+  './index.json',
+  './index.node',
+];
+
+const mainResolveReturnType = {
+  kInvalidUrl: -2,
+  kModuleNotFound: -1,
```

Comment:
I'm in favor of removing these error cases, and handling (throwing) them on C++ side.